### DISCUSSION TITLE
Create GenericKubernetesResource from file without ResourceDefinitionContext

### DISCUSSION THREAD

**Discussion by SoerenHenning** [fabric8io/kubernetes-client#3941] [Q&A] [answered]
Hi, I'm looking for a simple way to resemble something like `kubectl apply -f resource-of-any-kind.yaml` in Java. `resource-of-any-kind.yaml` might be a custom resource, which is not under my control so the [Typeless API](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md#resource-typeless-api) should be the way to go. However, to use that API I need to create a `ResourceDefinitionContext` object, which I do not really can since I don't know the resource type of `resource-of-any-kind.yaml` (only by looking into the file).

Is there any way to let fabric8 Kubernetes client create that context under the hood? If not, how can this be done programmatically? How does kubectl manage this?

**Comment by shawkins** [comment, marked as answer]
This is supported by 5.11+.  See the issues referenced with #3547.

**Comment by manusa** [comment]
Just to provide an example:

```java
kubernetesClient.resource("you-resource-as-yaml").createOrReplace();
kubernetesClient.load(yourResourceAsInputStream).createOrReplace();
```

The client will infer your context by querying the cluster for the APIs responsible of the provided resource Kind+apiVersion.

---

_Edit (2022-03-12):_

As suggested by @shawkins later [in this discussion](https://github.com/fabric8io/kubernetes-client/discussions/3941#discussioncomment-2339277), in case you are testing this implementation with our KubernetesMockServer, then you'd need to setup some expectations for your API (if your YAML contains an entry for a non-core resource):

```java
    ResourceDefinitionContext rdc = new ResourceDefinitionContext.Builder()
        .withGroup("test.fabric8.io")
        .withVersion("v1alpha1")
        .withKind("hello")
        .withNamespaced(true)
        .withPlural("hellos")
        .build();

    server.expect()
        .get()
        .withPath(String.format("/apis/%s/%s", rdc.getGroup(), rdc.getVersion()))
        .andReturn(HttpURLConnection.HTTP_OK,
            new APIResourceListBuilder()
                .withResources(
                    new APIResourceBuilder().withKind(rdc.getKind())
                        .withNamespaced(rdc.isNamespaceScoped())
                        .withName(rdc.getPlural())
                        .build())
                .build())
        .once();
```

**Reply by SoerenHenning** [reply]
Thank you and @shawkins. The clarification here helps a lot. Actually, I was looking for something more complicated after the previous reply. 😅 Maybe you can make this the selected answer.

**Reply by SoerenHenning** [reply]
Unfortunately this does not work, when running against the Kubernetes test server from kubernetes-server-mock. Running

```java
KubernetesServer server = new KubernetesServer(false, true);
server.client.load(serviceMonitorStream).get();
```

results in the following exception:

```
Could not find a registered handler for item: [GenericKubernetesResource(apiVersion=monitoring.coreos.com/v1, kind=ServiceMonitor, metadata=ObjectMeta(annotations=null, clusterName=null, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={...}, managedFields=[], name=test-service-monitor, namespace=null, ownerReferences=[], resourceVersion=null, selfLink=null, uid=null, additionalProperties={}), additionalProperties={})].
io.fabric8.kubernetes.client.KubernetesClientException: Could not find a registered handler for item: [GenericKubernetesResource(apiVersion=monitoring.coreos.com/v1, kind=ServiceMonitor, metadata=ObjectMeta(annotations=null, clusterName=null, creationTimestamp=null, deletionGracePeriodSeconds=null, deletionTimestamp=null, finalizers=[], generateName=null, generation=null, labels={...}, managedFields=[], name=test-service-monitor, namespace=null, ownerReferences=[], resourceVersion=null, selfLink=null, uid=null, additionalProperties={}), additionalProperties={})].
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.handlerOf(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:260)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.acceptVisitors(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableImpl.java:244)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.lambda$getItems$1(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:87)
	at java.base@11.0.2/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base@11.0.2/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base@11.0.2/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base@11.0.2/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base@11.0.2/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base@11.0.2/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base@11.0.2/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.getItems(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:89)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.getOperations(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:93)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.get(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:196)
	at io.fabric8.kubernetes.client.dsl.internal.NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.get(NamespaceVisitFromServerGetWatchDeleteRecreateWaitApplicableListImpl.java:54)
	...
```

Is this a known limitation or can we circumvent this? I'm using Quarkus Tests and Kubernete Client v5.12.1 if that matters.

**Reply by SoerenHenning** [reply]
Just saw #3960 now. That's the same problem, I am facing, right @shawkins?

------
*previous wrong comment deleted here*

------

Still, if loading generic Kubernetes resources via `kubernetesClient.load(...)` is not supported in the mock server right now, is there yet another way to test my code for Custom Resources, which is using `kubernetesClient.load(...)` under the hood?

**Reply by shawkins** [reply]
> That's the same problem, I am facing, right @shawkins?

I think so.  Implicit support will not currently work with the built-in mocking because the metadata isn't available from the mock server.  You'd have to create a specialized mock server that returns the necessary api group information, and I'm not sure how easy it is to then get that integrated with the quarkus annotations.  The alternative is at least for your tests use an explicit context.

**Reply by manusa** [reply]
A workaround for CRUD mode would be to add the specific expectations for the API resource metadata endpoints, but I haven't tested it.

Same expectations + others could be used for the non-CRUD mode.

**Reply by shawkins** [reply]
Yes that should work, no need to customize the server.  I was getting mixed up with how built-in handling was added for root paths.  You can use (and we can add a utility method to the KubernetesServer under #3960) in crud or non-crud:

```
    ResourceDefinitionContext rdc = new ResourceDefinitionContext.Builder()
        .withGroup("test.fabric8.io")
        .withVersion("v1alpha1")
        .withKind("hello")
        .withNamespaced(true)
        .withPlural("hellos")
        .build();

    server.expect()
        .get()
        .withPath(String.format("/apis/%s/%s", rdc.getGroup(), rdc.getVersion()))
        .andReturn(HttpURLConnection.HTTP_OK,
            new APIResourceListBuilder()
                .withResources(
                    new APIResourceBuilder().withKind(rdc.getKind())
                        .withNamespaced(rdc.isNamespaceScoped())
                        .withName(rdc.getPlural())
                        .build())
                .build())
        .once();
```

For this to work at a high level in crud mode we'd need to be able to issue an APIResource POST, but we don't have any method for that currently.

**Reply by manusa** [reply]
> For this to work at a high level in crud mode we'd need to be able to issue an APIResource POST, but we don't have any method for that currently.

Maybe special handling of APIService could take care of this, but it would really complicate things. 

Adding an accessor to be able to be able to manually register ResourceDefinitionContext in the KubernetesAttributesExtractor would be the easiest approach.

**Reply by shawkins** [reply]
Moved further discussion over to https://github.com/fabric8io/kubernetes-client/issues/3960#issuecomment-1065135838

**Reply by SoerenHenning** [reply]
Thank you for the huge support! I was able to get my tests run with the code snipped provided by @shawkins. With the help of Kotlin extension functions, this currently everything I need.

**Comment by manusa** [comment]
I'm going to edit my answer to include Steven's snippet so you can get a complete answer without going through the discussion.

**Comment by chenzhihao** [comment]
Thanks! This answer is pretty helpful. Could we add this to the document?
